amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java

```diff
@@ -299,6 +313,38 @@ private synchronized void assertHashRangeOfClosedShardIsCovered(final Shard clos
         return shardIdToChildShardIdsMap;
     }
 
+    /**
+     * Helper method to resolve the correct shard filter to use when listing shards from a position in a stream.
+     * @param initialPositionInStreamExtended
+     * @return
+     */
+    private static ShardFilter getShardFilterFromInitialPosition(InitialPositionInStreamExtended initialPositionInStreamExtended) {
+
+        ShardFilter.Builder builder = ShardFilter.builder();
+
+        switch (initialPositionInStreamExtended.getInitialPositionInStream()) {
+            case LATEST:
+                builder = builder.type(ShardFilterType.AT_LATEST);
+                break;
+            case TRIM_HORIZON:
+                builder = builder.type(ShardFilterType.AT_TRIM_HORIZON);
+                break;
+            case AT_TIMESTAMP:
+                builder = builder.type(ShardFilterType.AT_TIMESTAMP).timestamp(initialPositionInStreamExtended.getTimestamp().toInstant());
+                break;
+        }
+        return builder.build();
+    }
+
+    private static List<Shard> getShardListAtInitialPosition(@NonNull final ShardDetector shardDetector,
+                                                     InitialPositionInStreamExtended initialPositionInStreamExtended) throws KinesisClientLibIOException {
+        final ShardFilter shardFilter = getShardFilterFromInitialPosition(initialPositionInStreamExtended);
+        final Optional<List<Shard>> shards = Optional.of(shardDetector.listShardsWithFilter(shardFilter));
+
+        return shards.orElseThrow(() -> new KinesisClientLibIOException("Stream is not in ACTIVE OR UPDATING state - " +
```

Comment:
We need streamidentifier details in the exception.

Reply:
Sure, added.